### PR TITLE
Add functions to check Range, IP and CIDR. Included test and docs.

### PR DESCRIPTION
**Pull request by kemasdimas**
I accidentally closed #59 while cleaning up my forked repo 😓. All the commits have been squashed into a single commit for now.

### CODE REVIEW

**Comment by kemasdimas** [comment]
## IPv6 Consideration
My mind can't stop thinking about IPv6 implementation, so it's  better to write it here 😊

Looking back to the `eval` code
```
fun eval(formula: String, env: JsonObject): Float {
    try {
        return when (val result = RuleGrammar.parseToEnd(formula).eval(env)) {
            is Boolean -> if (result) 1f else 0f
            is Number -> result.toFloat()
            is String -> result.runCatching { toFloat() }.recoverCatching { 0f }.getOrThrow()
            else -> 0f
        }
...
}
```
If it was designed to convert any result to `Float` which is then be used in the `isEnabled` function. 

I think we should reconsider using a separate function to process both IPv4 and IPv6 in 1 function, and this new function will return either `1f` or `0f`

## Why?
Converting IPv6 can't be done with a single variable, but in this case, using an array is not possible either. Because the `eval` will always return `Float`.  

And currently, I can't find a way to accurately represent 2 `Long` variables into 1 `Float` 😁

So adding IPv6 support in the current `ip` function seems to be incompatible with `eval`

## Proposal
We can create one function to support both IPv4 and IPv6, i.e `ipmatches("2001:4860:4860::8888/32", "2001:4860:4860::8888")`

I know... we'll be back to the first proposal (creating a separate function just for IP)😅

But, this single function will accept both IP formats, without needing to revamp the whole `eval` concept

**Comment by goncalossilva** [comment]
About the PR itself, I left a couple of follow-ups on #59 since they were inline. Please have a look.

In general, I think we could merge IPv4 support and continue discussing IPv6 in a new ticket, but I'm fine to combine both together if you feel that's best. I'm enjoying thinking this through and ensuring we have a great design and implementation. Thanks for the back-and-forth!

> If it was designed to convert any result to Float which is then be used in the isEnabled function.

Something to note about `eval` is that it's the _last_ operation of all. There are intermediate representations of data (string, boolean, number, array, etc.) and that's OK. In fact, many functions expect to work with certain data structures that are not floats.

It's also worth noting that the existing functions -- `eq`, `lt` (+ all others), `contains` -- don't make many assumptions. As long as something understands equality and implements `Comparable`, it should work. For example, if `ipv6` returned an internal class backed by 2 longs that implements `Comparable` and `equals`/`hashCode`, and `cidrv6` returned a range between 2 of those instances, the rest should mostly work out of the box.

We could also look into supporting both formats under `ip` and `cidr`. I don't think it would require many changes, but to be fair, I haven't prototyped it.

> And currently, I can't find a way to accurately represent 2 Long variables into 1 Float 😁

It's also impossible to represent a single `Long` in a `Float`, since a few bits go into exponent precision. The max value is around 2^53, not 2^64! But hopefully, it's clear why that's not a problem -- converting to `Float` is the very last step, and we're actually just looking to get it within the `[0,1]` range. 😊

**Comment by kemasdimas** [comment]
I will address the comments in https://github.com/Doist/ffs/pull/59#discussion_r796191947 to finalize IPv4 implementation

And then I'll open a new issue on IPv6 to keep the discussion relevant 😊

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 1]
```diff
@@ -184,8 +192,16 @@ private sealed class RuleExpr<out T> {
...
+    data class ArrayExpr<out T>(val list: List<RuleExpr<T>>) : RuleExpr<Collection<T>>() {
+        override fun eval(env: JsonObject): Collection<T> = list.map { it.eval(env) }
+    }
+
+    data class RangeExpr<out T>(
+        val from: RuleExpr<T>,
+        val to: RuleExpr<T>
+    ) : RuleExpr<Collection<Long>>() {
+        override fun eval(env: JsonObject) =
+            delegateRangeToCollection(LongRange(from.eval(env) as Long, to.eval(env) as Long))
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]:
```suggestion
            delegateRangeToCollection(from.castEval<Long>(env)..to.castEval(env)))
```

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 2]
```diff
@@ -442,6 +458,60 @@ private sealed class RuleExpr<out T> {
...
+                    min[i] = castedValue and subnet[i]
+                    max[i] = castedValue or (subnet[i].inv())
+                }
+
+                return delegateRangeToCollection(calculateIpv4Value(min)..calculateIpv4Value(max))
+            }
+        }
+
+        companion object {
+            fun calculateIpv4Value(octets: List<UByte>): Long {
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]:
```suggestion
            fun sumOctets(octets: List<UByte>): Long {
```

Weak suggestion, but “calculate IPv4 value” is a bit ambiguous. 😄

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 3]
```diff
@@ -507,5 +579,17 @@ private sealed class RuleExpr<out T> {
         private fun Iterator<*>.fastForward() {
             while (hasNext()) next()
         }
+
+        @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
+        private fun delegateRangeToCollection(
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]:
```suggestion
        private fun wrapRangeInCollection(
```

Delegate has a “special” meaning in Kotlin, so I'd prefer we use an alternative term.

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 4]
```diff
@@ -226,6 +226,19 @@ class RuleEvalTest {
         assertFailsWith<IllegalArgumentException> { eval("""contains("+01:00", "+01:00")""") }
     }
 
+    @Test
+    fun range() {
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:231]: Let's add a couple of tests for edge cases:
- Inverted ranges (e.g., `[10:0]`)
- At the precision limit of Int/Long.

**Review by goncalossilva — COMMENTED** [review]
Left a few final comments, mostly about naming and missing test scenarios. There's also a failing test case:

`assertEquals(3232238083f, eval("""ip("192.168.10.3")"""))`

We should be good to merge after this! 😄

**Comment by kemasdimas** [comment]
> There's also a failing test case:
> 
> `assertEquals(3232238083f, eval("""ip("192.168.10.3")"""))`

Interesting... this test case was passing when I ran it from IntelliJ (by running the commonTest folder)

image

image

If you have alternative ways to run the test suites, please let me know. Would like to make sure that the implementation is bug free 😊

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt (hunk 4 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:231]: > Inverted ranges (e.g., [10:0])

About inverted ranges, what's the expected outcome of `contains[10:0, 8]`?

I think it's better to invalidate inverted ranges input and throw `IllegalArgumentException`, to keep the rule format consistent within the dashboard 😊

> At the precision limit of Int/Long.

Should we support negative value in the Range? e.g. [-1000, 140]. Right now I can't think of a common use case that will require a negative value range for the feature flag.

What do you think?

Comment (goncalossilva) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:231]: > `contains[10:0, 8]`

This syntax is incorrect, right?

> I think it's better to invalidate inverted ranges input and throw IllegalArgumentException, to keep the rule format consistent within the dashboard 😊

Throwing on inverted ranges sounds good. 👍

> Should we support negative value in the Range? e.g. [-1000, 140]. Right now I can't think of a common use case that will require a negative value range for the feature flag.

I can't either, but doesn't support come for free just by using signed types (the default)?

**Comment by goncalossilva** [comment]
The way CI runs tests is here: https://github.com/Doist/ffs/blob/9b1e61c9538708831155d3ea06efd70e0d8519a9/.github/workflows/ci.yml#L21

I see the failing test in the checks of this PR. Don't you?

**Comment by kemasdimas** [comment]
> The way CI runs tests is here:
> 
> https://github.com/Doist/ffs/blob/9b1e61c9538708831155d3ea06efd70e0d8519a9/.github/workflows/ci.yml#L21
> 
> I see the failing test in the checks of this PR. Don't you?

Wow, this one got me good! 😅 never should've carelessly converted `Long` to `Float` in the first place, let alone created a test for that

> It's also impossible to represent a single Long in a Float, since a few bits go into exponent precision. The max value is around 2^53, not 2^64!

Thank you for pointing this out!

**Comment by kemasdimas** [comment]
> The way CI runs tests is here:
> 
> https://github.com/Doist/ffs/blob/9b1e61c9538708831155d3ea06efd70e0d8519a9/.github/workflows/ci.yml#L21
> 
> I see the failing test in the checks of this PR. Don't you?

Wow, this one got me good! 😅 never should've carelessly converted `Long` to `Float` in the first place, let alone created a test for that. All failed tests came from the JS platform, I'll make sure to check tests on all platforms before committing

I should've checked this beforehand
image

> It's also impossible to represent a single Long in a Float, since a few bits go into exponent precision. The max value is around 2^53, not 2^64!

Thank you for pointing this out!

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 5]
```diff
@@ -113,8 +113,11 @@ private object RuleGrammar : Grammar<RuleExpr<*>>() {
             RuleExpr.NumberExpr(value)
         }
     private val long = optional(minus) and digits map { (minus, int) ->
-        var value = int.text.toLong()
-        minus?.let { value -= value }
+        val value = if (minus != null) {
+            "-${int.text}".toLong()
```
Comment (kemasdimas) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt:117]: This is needed to support the lowest Long value

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 6]
```diff
@@ -296,12 +304,6 @@ class RuleEvalTest {
 
     @Test
     fun ip() {
-        assertEquals(167772160f, eval("""ip("10.0.0.0")"""))
```
Comment (kemasdimas) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]: Removed all IP to `Float` conversion, it's not needed for the test because

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 7]
```diff
@@ -113,8 +113,11 @@ private object RuleGrammar : Grammar<RuleExpr<*>>() {
             RuleExpr.NumberExpr(value)
         }
     private val long = optional(minus) and digits map { (minus, int) ->
-        var value = int.text.toLong()
-        minus?.let { value -= value }
+        val value = if (minus != null) {
```
Comment (kemasdimas) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt:116]: This is needed to support the lowest Long value (-9223372036854775808)

Comment (goncalossilva) [reply ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt:116]: Oh, nice catch! That `value -= value` was meant to be `value = -value`, as it appears on the `Double` expression right above. It does reduce support back down to -9223372036854775807, but that feels like the right trade-off for consistency and simplicity.

**Comment by goncalossilva** [comment]
> All failed tests came from the JS platform, I'll make sure to check tests on all platforms before committing

Ah, yes, JS is tricky. The entire reason `castEval` exists is because of JS's leniency with types.

**Comment by goncalossilva** [comment]
CI is green, we're good to go! There are a few tiny changes I'd like to make, but I'll do them right after. The PR is gold, thanks for contributing!

**Comment by kemasdimas** [comment]
You're welcome, thanks for all the time you've set aside to review my pull request!

I've opened new issue #65 to start implementing IPv6

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 8]
```diff
@@ -281,6 +302,53 @@ class RuleEvalTest {
...
+        assertEquals(1f, eval("""contains(cidr("0.0.0.0/0"), ip("41.173.112.199"))"""))
+
+        assertEquals(0f, eval("""contains([cidr("254.200.222.210/23")], ip("254.200.222.25"))"""))
+        assertFailsWith<IllegalArgumentException> {
+            eval("""contains(cidr("254.200.222.210/23"))""")
+        }
+        assertFailsWith<IllegalArgumentException> {
+            eval("""contains([cidr("254.200.222.210/23", "254.200.224.0")])""")
+        }
+    }
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:350]: I've just realized that both `cird()` and `ip()` tests have somewhat of a large scope. Not only do they test the respective functions, they also test `contains` and `gt` (and friends). It would be great to make them more focused by testing specifically the outputs of each, and delegating a few of the ones that compose with other function to `composition`.

Comment (kemasdimas) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:350]: I'm thinking of changing `private sealed class RuleExpr<out T>` to public, this will help us test the IP classes without having to do composition

Currently, the `ip` and `cidr` has no exposed public method, so either 
1. We made it public, or
2. Create a test helper method within RuleEval.kt

Alternative 1 is cleaner in my opinion

Comment (goncalossilva) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:350]: What is the blocker to testing via eval? Precision?

Comment (kemasdimas) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:350]: Precision is one of the considerations, but the bigger blocker to testing is the `cidr` returns range, which can't be converted with `eval`. This is the reason I'm using `contains` to run the test

This challenge is even more prominent on IPv6 implementation proposal

```kotlin
return when (val result = RuleGrammar.parseToEnd(formula).eval(env)) {
    is Boolean -> if (result) 1f else 0f
    is Number -> result.toFloat()
    is String -> result.runCatching { toFloat() }.recoverCatching { 0f }.getOrThrow()
    else -> 0f
}
```

This got me thinking, what about extracting all the IP related functions to a data class, and testing that class separately outside RuleEval

It might be a bit excessive, but it will decouple RuleEval logic from IP related logic

Comment (goncalossilva) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:350]: Gotcha. Let's keep the logic within that file as it's all connected and only used there, but the approach seems sound! That would benefit most other tests, too. One suggestion is to use `internal` instead of `public` since these modules are not meant to be used externally.

Comment (kemasdimas) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:350]: Sounds good! 😊👍
